### ISSUE TITLE
bug: steam IO encode performance is unexpect,  freeBytes is not done

### ISSUE THREAD

**Issue by 201430098137** [bytedance/sonic#542]
recently, I use stram IO to reduct memory usage, but memory usage is higher then sonic.Marshal, I bench it.
I encode map[string]interface{} to json and write it to file,
```
var cases = []map[string]interface{}{
	{
		"dsds": "dsdds",
		"dsalkjlsd":   "sasasas",
		"dsalkjlsdd":  "sasasas",
		"dsalkjlsds":  "sasasas",
		"dsalkjlasd":  "sasasas",
		"dsalkjlsdsf": "sasasas",
	},
	{
		"dsdsd": "dsdsd",
		"dsalkjalsd":   "sasasas",
		"dsalkjlsdsdd":  "sasasas",
		"dsalkjlsdss":  "sasasas",
		"dsalkjlasdsd":  "sasasas",
		"dsalkjlssfddsf": "sasasas",
		"dsalkjlssdfd":   "sasasas",
		"dsalkjlssdd":  "sasasas",
		"dsalkjlssds":  "sasasas",
		"dsalkjlassd":  "sasasas",
		"dsalkjlscdsf": "sasasas",
		"dsalkjlsdsd":   "sasasas",
		"dsalkjlsdd":  "sasasas",
		"dsalkjlsds":  "sasasas",
		"dsalkjlasd":  "sasasas",
		"dsalkjlsdsf": "sasasas",
	},
	{
		"dsdsd": "dsdsd",
		"dsalkjalsd":   "sasasas",
		"dsalkjlsdsdd":  "sasasas",
		"dsalkjlsdss":  "sasasas",
		"dsalkjlasdsd":  "sasasas",
		"dsalkjlssfddsf": "sasasas",
		"dsalkjlssdfd":   "sasasas",
		"dsalkjlssdd":  "sasasas",
		"dsalkjlssds":  "sasasas",
		"dsalkjlassd":  "sasasas",
		"dsalkjlscdsf": "sasasas",
		"dsalkjlsdsd":   "sasasas",
		"dsalkjlsdd":  "sasasas",
		"dsalkjlsds":  "sasasas",
		"dsalkjlasd":  "sasasas",
		"dsalkjlsdsf": "sasasas",
		"sasdas":"asa",
		"saasdss":"asa",
		"sassdc":"asa",
		"sasvsdsdas":"asa",
		"sasjyja":"asa",
		"sagncs":"asa",
		"sabdfss":"asa",
		"santys":"asa",
		"saabfdss":"asa",
		"sasbdfc":"asa",
		"sasbdfas":"asa",
		"sabdfsa":"asa",
		"sacdbfas":"asa",
		"sasss":"asa",
		"sasaasf":"asa",
		"saasdass":"asa",
		"sasavsasc":"asa",
		"sfda":"asa",
		"saasfsa":"asa",
		"sasafscs":"asa",
		"asf":"asa",
		"vxxz":"asa",
		"zxc":"asa",
		"nrt":"asa",
		"dbf":"asa",
		"sabwdfsa":"asa",
		"bdf":"asa",
		"sazxvss":"asa",
		"sasvzv":"asa",
		"saaczczxss":map[string]interface{}{

			"dsdsd": "dsdsd",
			"dsalkjalsd":   "sasasas",
			"dsalkjlsdsdd":  "sasasas",
			"dsalkjlsdss":  "sasasas",
			"dsalkjlasdsd":  "sasasas",
			"dsalkjlssfddsf": "sasasas",
			"dsalkjlssdfd":   "sasasas",
			"dsalkjlssdd":  "sasasas",
			"dsalkjlssds":  "sasasas",
			"dsalkjlassd":  "sasasas",
			"dsalkjlscdsf": "sasasas",
			"dsalkjlsdsd":   "sasasas",
			"dsalkjlsdd":  "sasasas",
			"dsalkjlsds":  "sasasas",
			"dsalkjlasd":  "sasasas",
			"dsalkjlsdsf": "sasasas",
		},
		"cz":"asa",
		"sasvzxas":map[string]interface{}{

			"dsdsd": "dsdsd",
			"dsalkjalsd":   "sasasas",
			"dsalkjlsdsdd":  "sasasas",
			"dsalkjlsdss":  "sasasas",
			"dsalkjlasdsd":  "sasasas",
			"dsalkjlssfddsf": "sasasas",
			"dsalkjlssdfd":   "sasasas",
			"dsalkjlssdd":  "sasasas",
			"dsalkjlssds":  "sasasas",
			"dsalkjlassd":  "sasasas",
			"dsalkjlscdsf": "sasasas",
			"dsalkjlsdsd":   "sasasas",
			"dsalkjlsdd":  "sasasas",
			"dsalkjlsds":  "sasasas",
			"dsalkjlasd":  "sasasas",
			"dsalkjlsdsf": "sasasas",
		},
		"vz":map[string]interface{}{

			"dsdsd": "dsdsd",
			"dsalkjalsd":   "sasasas",
			"dsalkjlsdsdd":  "sasasas",
			"dsalkjlsdss":  "sasasas",
			"dsalkjlasdsd":  "sasasas",
			"dsalkjlssfddsf": "sasasas",
			"dsalkjlssdfd":   "sasasas",
			"dsalkjlssdd":  "sasasas",
			"dsalkjlssds":  "sasasas",
			"dsalkjlassd":  "sasasas",
			"dsalkjlscdsf": "sasasas",
			"dsalkjlsdsd":   "sasasas",
			"dsalkjlsdd":  "sasasas",
			"dsalkjlsds":  "sasasas",
			"dsalkjlasd":  "sasasas",
			"dsalkjlsdsf": "sasasas",
		},
		"sazxvcs":map[string]interface{}{

			"dsdsd": "dsdsd",
			"dsalkjalsd":   "sasasas",
			"dsalkjlsdsdd":  "sasasas",
			"dsalkjlsdss":  "sasasas",
			"dsalkjlasdsd":  "sasasas",
			"dsalkjlssfddsf": "sasasas",
			"dsalkjlssdfd":   "sasasas",
			"dsalkjlssdd":  "sasasas",
			"dsalkjlssds":  "sasasas",
			"dsalkjlassd":  "sasasas",
			"dsalkjlscdsf": "sasasas",
			"dsalkjlsdsd":   "sasasas",
			"dsalkjlsdd":  "sasasas",
			"dsalkjlsds":  "sasasas",
			"dsalkjlasd":  "sasasas",
			"dsalkjlsdsf": "sasasas",
		},
		"sasvzxs":map[string]interface{}{

			"dsdsd": "dsdsd",
			"dsalkjalsd":   "sasasas",
			"dsalkjlsdsdd":  "sasasas",
			"dsalkjlsdss":  "sasasas",
			"dsalkjlasdsd":  "sasasas",
			"dsalkjlssfddsf": "sasasas",
			"dsalkjlssdfd":   "sasasas",
			"dsalkjlssdd":  "sasasas",
			"dsalkjlssds":  "sasasas",
			"dsalkjlassd":  "sasasas",
			"dsalkjlscdsf": "sasasas",
			"dsalkjlsdsd":   "sasasas",
			"dsalkjlsdd":  "sasasas",
			"dsalkjlsds":  "sasasas",
			"dsalkjlasd":  "sasasas",
			"dsalkjlsdsf": "sasasas",
		},
		"sl;kas":map[string]interface{}{

				"dsdsd": "dsdsd",
				"dsalkjalsd":   "sasasas",
				"dsalkjlsdsdd":  "sasasas",
				"dsalkjlsdss":  "sasasas",
				"dsalkjlasdsd":  "sasasas",
				"dsalkjlssfddsf": "sasasas",
				"dsalkjlssdfd":   "sasasas",
				"dsalkjlssdd":  "sasasas",
				"dsalkjlssds":  "sasasas",
				"dsalkjlassd":  "sasasas",
				"dsalkjlscdsf": "sasasas",
				"dsalkjlsdsd":   "sasasas",
				"dsalkjlsdd":  "sasasas",
				"dsalkjlsds":  "sasasas",
				"dsalkjlasd":  "sasasas",
				"dsalkjlsdsf": "sasasas",
		},

	},
}
func BenchmarkEncode(b *testing.B) {
	fd, err := os.OpenFile("/tmp/aa.json", os.O_RDWR|os.O_CREATE|os.O_APPEND, os.ModePerm)
	if err != nil {
		b.Errorf("打开文件错误：%s", err.Error())
	}
	var enc = sonic.ConfigDefault.NewEncoder(fd)
	for i := 0; i < b.N; i++ {
		err := enc.Encode(cases[i%len(cases)])
		if err != nil {
			b.Errorf("enc.Encode错误：%s", err.Error())
		}
	}
}

func BenchmarkMarshal(b *testing.B) {
	fd, err := os.OpenFile("/tmp/bb.json", os.O_RDWR|os.O_CREATE|os.O_APPEND, os.ModePerm)
	if err != nil {
		b.Errorf("打开文件错误：%s", err.Error())
	}

	for i := 0; i < b.N; i++ {
		bytes, err := sonic.Marshal(cases[i%len(cases)])
		if err != nil {
			b.Errorf("sonic.Marshal错误：%s", err.Error())
		}
		_, err = fd.Write(bytes)
		if err != nil {
			b.Errorf("Write错误：%s", err.Error())
		}
		_, err = fd.Write([]byte{'\n'})
		if err != nil {
			b.Errorf("Write错误：%s", err.Error())
		}
	}

}

```
profile memory
stream IO
![image](https://github.com/bytedance/sonic/assets/13937318/26d7851c-8bb6-4f95-a1f3-ad13a0658a62)
sonic.Marshal
![image](https://github.com/bytedance/sonic/assets/13937318/20a98484-3300-4522-8e75-abea924d7614)

Interestingly, stream IO use more memory then sonic.Marshal.
I debugger that code, I discover the issue.

![image](https://github.com/bytedance/sonic/assets/13937318/637efb6b-476c-477e-80ef-3c313e297b05)
line 71 out is set out[n:], data of out is unaccess, line 81 freeBytes is useless.  
![image](https://github.com/bytedance/sonic/assets/13937318/1e7fb1e9-5b4c-48eb-9806-9c17d5370e33)

**Comment by 201430098137** [comment]
code:  github.com/bytedance/sonic@v1.10.2/internal/encoder/stream.go:70

**Comment by liuq19** [comment]
Thanks. we will investigate it.  cc @AsterDY

**Comment by AsterDY** [comment]
fixed by #547
